server/src/main/java/org/elasticsearch/common/blobstore/RetryingInputStream.java

```diff
@@ -0,0 +1,329 @@
+                failuresAfterMeaningfulProgress,
+                maxRetriesForNoMeaningfulProgress()
+            ),
+            e
+        );
+    }
+
+    private void maybeLogAndRecordMetricsForSuccess(int initialAttempt, String action) {
+        if (attempt > initialAttempt) {
+            final int numberOfRetries = attempt - initialAttempt;
+            logger.info(
+                "successfully {} input stream for [{}] with purpose [{}] after [{}] retries",
+                action,
+                blobStoreServices.getBlobDescription(),
+                purpose.getKey(),
+                numberOfRetries
+            );
+            blobStoreServices.onRetrySucceeded(action, numberOfRetries);
+        }
+    }
+
+    private long currentStreamProgress() {
+        return Math.subtractExact(Math.addExact(start, currentOffset), currentStreamFirstOffset);
+    }
+
+    private boolean shouldRetry(int attempt) {
+        if (purpose == OperationPurpose.REPOSITORY_ANALYSIS) {
+            return false;
+        }
+        if (purpose == OperationPurpose.INDICES) {
+            return true;
+        }
+        final int maxAttempts = blobStoreServices.getMaxRetries() + 1;
+        return attempt < maxAttempts + failuresAfterMeaningfulProgress;
+    }
+
+    private int maxRetriesForNoMeaningfulProgress() {
+        return purpose == OperationPurpose.INDICES ? Integer.MAX_VALUE : (blobStoreServices.getMaxRetries() + 1);
+    }
+
+    private void delayBeforeRetry(long delayInMillis) {
+        try {
+            assert shouldRetry(attempt - 1) : "should not have retried";
+            Thread.sleep(delayInMillis);
+        } catch (InterruptedException e) {
+            logger.info("s3 input stream delay interrupted", e);
+            Thread.currentThread().interrupt();
+        }
+    }
+
+    // protected access for testing
+    protected long getRetryDelayInMillis() {
+        // Initial delay is 10 ms and cap max delay at 10 * 1024 millis, i.e. it retries every ~10 seconds at a minimum
+        return 10L << (Math.min(attempt - 1, 10));
+    }
+
+    @Override
+    public void close() throws IOException {
+        try {
+            currentStream.inputStream.close();
+        } finally {
+            closed = true;
+        }
+    }
+
+    @Override
+    public long skip(long n) throws IOException {
+        ensureOpen();
+        return currentStream.inputStream.skip(n);
+    }
+
+    @Override
+    public void reset() {
+        throw new UnsupportedOperationException("RetryingInputStream does not support seeking");
+    }
+
+    private <T extends Exception> T addSuppressedExceptions(T e) {
+        for (Exception failure : failures) {
+            e.addSuppressed(failure);
+        }
+        return e;
+    }
+
+    /**
+     * This implements all the behavior that is blob-store-specific
+     *
+     * @param <V> The type of the version used
+     */
+    protected interface BlobStoreServices<V> {
```

Comment:
I parameterised this because I thought we might have different representations of versions for each object store, but I think it could probably just be a string because they seem to all support `ifMatch = <etag>`